databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultDatabus.java

```diff
@@ -740,7 +790,109 @@ private void checkSubscriptionOwner(String ownerId, OwnedSubscription subscripti
         }
     }
 
-    /** EventStore sink that doesn't count adjacent events for the same table/key against the peek/poll limit. */
+
+    private void drainQueueAsync(String subscription) {
+        try {
+            boolean notDrainingNow = _drainedSubscriptionsMap.putIfAbsent(subscription, Boolean.TRUE) == null;
+            if (notDrainingNow) {
+                _log.info("Starting the draining process for subscription: {}.", subscription);
+                _drainService.submit(new Runnable() {
+                    @Override
+                    public void run() {
+                        try {
+                            doDrainQueue(subscription);
+                        } finally {
+                            _drainedSubscriptionsMap.remove(subscription);
+                        }
+                    }
+                });
+            } else {
+                _log.info("Draining for subscription: {} is already in progress.", subscription);
+            }
+        } catch (Exception e) {
+            Throwables.propagate(e);
+        }
+    }
+
+    private void doDrainQueue(String subscription) {
+
+        boolean nonRedundantItemFound = false;
+
+        // this loop will break once the peek found a non-redundant item or when the queue gets empty.
+        while (true) {
```

Comment:
Instead of using a while loop I'd recommend re-submitting the task back onto `_drainService`.  As currently written once there are 10 subscriptions in need of draining the 11th onward are blocked until one of the others finish.  By re-submitting on each iteration it provides some level of fairness, or at least minimizes the chances of starvation.